src/richtext/utils/RichTextFieldBuilder.ts

```diff
@@ -0,0 +1,159 @@
+ *
+ * Taken from the `type-fest` package.
+ *
+ * @see https://github.com/sindresorhus/type-fest/blob/8a45ba048767aaffcebc7d190172d814a739feb0/source/distributed-omit.d.ts
+ */
+type DistributedOmit<
+	ObjectType,
+	KeyType extends ObjectType extends unknown ? keyof ObjectType : never,
+> = ObjectType extends unknown ? Omit<ObjectType, KeyType> : never;
+
+/**
+ * An inline node without its `start` and `end` properties.
+ */
+export type RTPartialInlineNode = DistributedOmit<
+	RTInlineNode,
+	"start" | "end"
+>;
+
+export class RichTextFieldBuilder {
+	private nodes: RTNode[] = [];
+
+	private get tail(): RTNode | undefined {
+		return this.nodes[this.nodes.length - 1];
+	}
+
+	appendNode(node: RTNode): this {
+		this.cleanupTail();
+
+		this.nodes.push(node);
+
+		return this;
+	}
+
+	appendTextNode(type: RTTextNode["type"], direction?: "ltr" | "rtl"): this {
+		return this.appendNode({
+			type: type,
+			text: "",
+			spans: [],
+			direction: direction || "ltr",
+		});
+	}
+
+	appendSpan(span: RTInlineNode): this {
+		if (!this.isOfKindText(this.tail)) {
+			throw new Error("Cannot add span to non-text tail node");
+		}
+
+		let lastIdenticalSpanIndex = -1;
+		for (let i = this.tail.spans.length - 1; i >= 0; i--) {
+			const lastSpan = this.tail.spans[i];
+
+			if (span.type === "strong" || span.type === "em") {
+				if (lastSpan.type === span.type) {
+					lastIdenticalSpanIndex = i;
+					break;
+				}
+			} else if (span.type === "label") {
+				if (
+					lastSpan.type === "label" &&
+					lastSpan.data.label === span.data.label
+				) {
+					lastIdenticalSpanIndex = i;
+					break;
+				}
+			} else if (span.type === "hyperlink") {
+				if (lastSpan.type === "hyperlink") {
+					// Check that all data fields are the same. Since it's not a
+					// deep equal operation, this doesn't guarantee a perfect
+					// match with content relationship.
+					const isSameLink = (
+						Object.entries(span.data) as [keyof RTLinkNode["data"], unknown][]
+					).every(([key, value]) => lastSpan.data[key] === value);
+
+					if (isSameLink) {
+						lastIdenticalSpanIndex = i;
+						break;
+					}
+				}
+			}
+		}
+
+		// Prefer to extend the last span of the same type end
+		// position if it ends at the start of the new span.
+		if (this.tail.spans[lastIdenticalSpanIndex]?.end === span.start) {
+			this.tail.spans[lastIdenticalSpanIndex].end = span.end;
+		} else {
+			this.tail.spans.push(span);
+		}
+
+		return this;
+	}
+
+	appendSpanOfLength(partialSpan: RTPartialInlineNode, length: number): this {
+		if (!this.isOfKindText(this.tail)) {
+			throw new Error("Cannot add span to non-text tail node");
+		}
+
+		const tailLenght = this.tail.text.length;
+
+		return this.appendSpan({
+			...partialSpan,
+			start: tailLenght,
+			end: tailLenght + length,
+		});
+	}
+
+	appendText(text: string): this {
+		if (!this.isOfKindText(this.tail)) {
+			throw new Error("Cannot append text to non-text tail node");
```

Comment:
I'm happy to do it, but `PrismicError` requires a URL and a response object. Do we want to make both optional now? Or do you have something else in mind? 🤔

Reply:
Ah, good point. I forgot `PrismicError` is designed as a fallback error for `fetch` calls. It would be better to make `PrismicError` a simple extension of `Error` and create a new `PrismicNetworkError` or something, but that would be a breaking change.

Perhaps we could use a new `PrismicRichTextError` in this case. We can refactor that error to extend `PrismicError` in the future when it is simplified.

Reply:
Refactored!